### PR TITLE
git add

### PR DESCRIPTION
**Pull request by KostiantynYehorov**
dev
## JIRA

* [Main JIRA ticket](https://jira.softserve.academy/secure/RapidBoard.jspa?rapidView=3191&projectKey=CPPLV&view=planning&selectedIssue=CPPLV-93)

## Code reviewers

CharlesLyleLaRue
Hipst0man
MaxAndrusenko
OksanaBisyk

### Second Level Review

StepBura

## Summary of issue

Add SMTP Server

## CHECK LIST
- [x]  СI passed
- [x]  Сode coverage >=95%
- [ ]  PR is reviewed manually again (to make sure you have 100% ready code)
- [ ]  All reviewers agreed to merge the PR
- [ ]  I've checked new feature as logged in and logged out user if needed
- [ ]  PR meets all conventions

### CODE REVIEW

File: SMTPServer/MailInfo.cpp [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+}
+
+const char* MailInfo::get_text() const
+{
+	return m_text.c_str();
+}
+
+void MailInfo::SaveToFile()
+{
+	std::ofstream file("email.txt", std::ofstream::app);
```
Comment (CharlesLyleLaRue) [inline SMTPServer/MailInfo.cpp:49]: Need a check if file is open.

File: SMTPServer/MailSession.h [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+#pragma comment(lib, "Ws2_32.lib")
+
+#define MAX_ADDRESS_LENGTH 256
+#define SMTP_DATA_TERMINATOR "\r\n.\r\n"
+#define FIRST_FOUR_SYMBOLS 4
+#define FIRST_EIGHT_SYMBOLS 8
+
+//#define SMTP_DATA_TERMINATOR "."
+
+enum MailSessionStatus
```
Comment (CharlesLyleLaRue) [inline SMTPServer/MailSession.h:20]: Consider to use enum class instead of plain enum.

File: SMTPServer/MailSession.h [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+	EMPTY,
+	EHLO,
+	MAIL_FROM,
+	RCPT_TO,
+	DATA,
+	SUBJECT,
+	QUIT
+};
+
+enum Responses
```
Comment (CharlesLyleLaRue) [inline SMTPServer/MailSession.h:34]: Consider to use enum class instead of plain enum.

File: SMTPServer/MailSession.h [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+	USER_NOT_LOCAL
+};
+
+class MailSession
+{
+public:
+	MailSession() = delete;
+	MailSession(SOCKET& client_socket);
+
+	const SOCKET& GetSocket() const;
```
Comment (CharlesLyleLaRue) [inline SMTPServer/MailSession.h]: Getter method should be like "get_socket" following our code convention.

File: SMTPServer/MailSession.cpp [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+bool MailSession::ValidAdress(char* buf)
+{
+	size_t strlen_buf = strlen(buf);
+	return (strlen_buf > 2 && strlen_buf < 255 && strchr(buf, '@'));
+}
+
+std::string MailSession::CutAddress(char* buf)
+{
+	std::string str_buf = buf;
+	std::string result = "";
```
Comment (CharlesLyleLaRue) [inline SMTPServer/MailSession.cpp:26]: Should we initialize string values?

File: SMTPServer/MailSession.cpp [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+		result += str_buf.at(i);
+	}
+
+	return result;
+}
+
+std::string MailSession::CutSubject(char* buf)
+{
+	std::string str_buf = buf;
+	std::string result = "";
```
Comment (CharlesLyleLaRue) [inline SMTPServer/MailSession.cpp:42]: String initialization.

File: SMTPServer/SMTPServer.cpp [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+	if (m_server_socket == INVALID_SOCKET)
+	{
+		std::cout << "Error with socket initializing!\n";
+		return false;
+	}
+
+	ZeroMemory(&server_info, sizeof(server_info));
+
+	server_info.sin_family = AF_INET;
+	server_info.sin_port = htons(25);
```
Comment (CharlesLyleLaRue) [inline SMTPServer/SMTPServer.cpp]: 25 is magic number here. Create an constexpr for default port.
constexpr int DEFAULT_PORT = "25";

File: SMTPServer/ThreadPool.h [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    void set_size(int thread_size);
+
+private:
+    void Join();
+    void Stop();
+    void DoTask();
+
+    bool IsStopped() { return is_stop; };
+
+private:
```
Comment (CharlesLyleLaRue) [inline SMTPServer/ThreadPool.h:40]: Private members should be named like this:
m_thread_size;
m_is_stop;

File: SMTPServer/ThreadPool.cpp [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+    {
+        Stop();
+    }
+}
+
+void ThreadPool::Join()
+{
+    for (int i = 0; i < thread_size; ++i)
+    {
+        m_thread_pool.at(i).join();
```
Comment (CharlesLyleLaRue) [inline SMTPServer/ThreadPool.cpp]: First check if thread is joinable, then join.

**Review by CharlesLyleLaRue — CHANGES_REQUESTED** [review]
Check my comments.

File: SMTPServer/MailSession.h (hunk 3 above)
Comment (KostiantynYehorov) [reply SMTPServer/MailSession.h:34]: I don't need to change it, bacause I often cast my enum types to int and I compare to different enums.

File: SMTPServer/MailSession.cpp (hunk 6 above)
Comment (KostiantynYehorov) [reply SMTPServer/MailSession.cpp:42]: I can't use "append()". It calls an exception

File: SMTPServer/MailSession.h (hunk 3 above)
Comment (CharlesLyleLaRue) [reply SMTPServer/MailSession.h:34]: You can cast enum class types to int as well.

**Review by CharlesLyleLaRue — DISMISSED** [review]
Looks good.

File: SMTPServer/MailInfo.cpp [hunk 10]
```diff
@@ -0,0 +1,54 @@
+// This is an independent project of an individual developer. Dear PVS-Studio, please check it.
+
+// PVS-Studio Static Code Analyzer for C, C++, C#, and Java: http://www.viva64.com
+
+#include "MailInfo.h"
+
+void MailInfo::set_mail_from(std::string& mail_from)
```
Comment (OksanaBisyk-TomTom) [inline SMTPServer/MailInfo.cpp]: If we get arguments by  reference and will not change it, we can write "const std::string& mail_from".

**Review by OksanaBisyk-TomTom — DISMISSED** [review]
Looks good.

**Review by MaxAndrusenko — DISMISSED** [review]
Cool

**Review by OksanaBisyk-TomTom — DISMISSED** [review]

**Review by CharlesLyleLaRue — DISMISSED** [review]

**Review by Globalika — DISMISSED** [review]

**Comment by Hipst0man** [comment]
LGTM

**Review by Hipst0man — DISMISSED** [review]
LGTM

**Review by MaxAndrusenko — APPROVED** [review]

**Review by Hipst0man — APPROVED** [review]
LGTM

**Review by CharlesLyleLaRue — APPROVED** [review]

**Pull request merged**